File: platforms/android/keyboard-renderer/src/main/java/app/funput/funput/keyboard/model/KeyboardEditorMode.kt
package app.funput.funput.keyboard.model

/** Describes the focused editor so the renderer and IME can apply matching behavior. */
enum class KeyboardEditorMode(
    val supportsVietnameseComposition: Boolean,
    val isNumber: Boolean = false,
    val isPassword: Boolean = false,
    val allowsDecimal: Boolean = false,
    val allowsSigned: Boolean = false,
) {
    TEXT(supportsVietnameseComposition = true),
    EMAIL(supportsVietnameseComposition = false),
    URL(supportsVietnameseComposition = false),
    PHONE(supportsVietnameseComposition = false),
    PASSWORD(supportsVietnameseComposition = false, isPassword = true),
    PIN(supportsVietnameseComposition = false, isPassword = true),
    NUMBER(supportsVietnameseComposition = false, isNumber = true),
    NUMBER_DECIMAL(supportsVietnameseComposition = false, isNumber = true, allowsDecimal = true),
    NUMBER_SIGNED(supportsVietnameseComposition = false, isNumber = true, allowsSigned = true),
    NUMBER_SIGNED_DECIMAL(
        supportsVietnameseComposition = false,
        isNumber = true,
        allowsDecimal = true,
        allowsSigned = true,
    ),
    ;

    val usesKeypad: Boolean get() = isNumber || this == PHONE || this == PIN
}

File: platforms/ios/Packages/FunputKit/Sources/KeyboardLayout/Models/KeyboardEditorMode.swift
public enum KeyboardEditorMode: String, CaseIterable, Hashable, Sendable {
    case text
    case search
    case email
    case url
    case phone
    case password
    case pin
    case number
    case numberDecimal
    case numberSigned
    case numberSignedDecimal

    public var supportsVietnameseComposition: Bool {
        self == .text || self == .search
    }

    public var isNumber: Bool {
        switch self {
        case .number, .numberDecimal, .numberSigned, .numberSignedDecimal: true
        default: false
        }
    }

    public var isPassword: Bool { self == .password || self == .pin }

    /// Whether the field is one a sentence is ever typed into.
    ///
    /// False for passwords, addresses and number pads, which is what keeps Funput's
    /// sentence detection out of them now that the user's preference can outrank a
    /// field asking for no capitalization at all.
    public var allowsAutoCapitalization: Bool { self == .text || self == .search }
    public var allowsDecimal: Bool { self == .numberDecimal || self == .numberSignedDecimal }
    public var allowsSigned: Bool { self == .numberSigned || self == .numberSignedDecimal }
    public var usesKeypad: Bool { isNumber || self == .phone || self == .pin }

    /// Whether this page can hide its number row and carry the digits as long-press
    /// alternates on the top character row instead. See `usesCompactLetterRows`.
    public var supportsCompactLetterRows: Bool {
        switch self {
        case .text, .search, .email, .url: true
        default: false
        }
    }

    /// Whether `KeyboardLayoutPreset.system` describes this mode's keyboard.
    ///
    /// The stock keyboard renders text and search identically — the magnifying glass on
    /// the return key comes from the enter action, not the layout. Every other mode is
    /// left to the Funput preset. Deliberately not folded into
    /// `supportsVietnameseComposition`, which happens to cover the same two cases today
    /// but answers a different question and may not track this one.
    public var usesSystemPreset: Bool { self == .text || self == .search }
}
